src/gpu_sysman.c

```diff
@@ -591,19 +613,37 @@ static int gpu_init(void) {
   return gpu_config_init(count);
 }
 
-/* Dispatch given value to collectd */
-static void gpu_submit(const char *name, const char *type, const char *valname,
-                       double value) {
-  value_list_t vl = VALUE_LIST_INIT;
-  vl.values = &(value_t){.gauge = value};
-  vl.values_len = 1;
-
-  sstrncpy(vl.plugin, PLUGIN_NAME, sizeof(vl.plugin));
-  sstrncpy(vl.plugin_instance, name, sizeof(vl.plugin_instance));
-  sstrncpy(vl.type_instance, valname, sizeof(vl.type_instance));
-  sstrncpy(vl.type, type, sizeof(vl.type));
+/* Dispatch given value to collectd.  Resets metric family after dispatch */
+static void gpu_submit(gpu_device_t *gpu, metric_family_t *fam) {
+  metric_t *m = fam->metric.ptr;
+  for (size_t i = 0; i < fam->metric.num; i++) {
+    metric_label_set(m + i, "pci_bdf", gpu->pci_bdf);
+  }
+  int status = plugin_dispatch_metric_family(fam);
+  if (status != 0) {
+    ERROR(PLUGIN_NAME ": gpu_submit(%s, %s) failed: %s", gpu->pci_bdf,
+          fam->name, strerror(status));
+  }
+  metric_family_metric_reset(fam);
+}
 
-  plugin_dispatch_values(&vl);
+/* because of family name change, each RAS metric needs to be submitted +
+ * reseted separately */
+static void ras_submit(gpu_device_t *gpu, const char *name, const char *help,
+                       const char *type, const char *subdev, double value) {
+  metric_family_t fam = {
+      .type = METRIC_TYPE_COUNTER,
+      .name = (char *)name,
```

Comment:
I hope these don't get written to, as in case if the incoming is truly const, writes result in undefined behavior

Reply:
Compilers puts all literal strings to read-only section by default, so in actuality they are const strings.  Assigning them to const char*, and propagating them e.g. to ras_submit() as such, is the correct thing to do, regardless of the enabled compiler warnings [1].

The annoying thing here is that metric_family_t.name in collectd v6 is non-const string.  This is because collectd offers also functions for allocation and freeing metric families, and for this purpose the .name member cannot be static.

My code does not use / need the metric family alloc / free functions, only the metrics processing ones, explicitly so casting the literal name + help strings to non-const is safe.  Same cast happens implicitly [1] in all the places where literal strings are assigned directly to metric_family_t.name in the code.

[1] GCC does not do const checks for literal strings even with "-Wall -Wextra", one needs to explicitly use "-Write-strings" before GCC emits warnings about their assignments.  "-Write-strings" GCC option would cause a large number of warnings about "metric_family_t.name" usage elsewhere than in this function, *and* in all the other collectd plugins though.

Allocating all the name & help strings just to be able to enable "-Wwrite-strings" warning is not worth the overhead though.